### DISCUSSION TITLE
Catching a custom event and running JavaScript after DOM re-rendering

### DISCUSSION THREAD

**Discussion by fabiomlferreira** [livewire/livewire#8578] [1. Help]
I'm firing a custom event on the backend and catching it on the frontend using Livewire.on. This event carries data that I need to create a new graph. However, the Livewire.on listener is triggered before the DOM is updated, so the element where I want to display the graph isn't available yet.

While using setTimeout makes it work functionally, it's not an ideal solution.

Is there a better way to catch a custom event and ensure the associated JavaScript runs only after the page content has been re-rendered?

**Comment by aoxborrow** [comment]
This is the same as setTimeout but I use a generic `rendered` event on all my components, then I use [$nextTick()](https://alpinejs.dev/magics/nextTick) in Alpine init(). There's probably a better way using the [morph hooks](https://livewire.laravel.com/docs/javascript#dom-morph-hooks), but feels like `$wire.$rendered(() => {...})` should be built-in...

```php
public function rendered() {
    // event so JS components can listen for re-render
    $this->dispatch('rendered');
}
```

```js
init() {
    // listen for renders and trigger onRender
    this.$wire.on('rendered', () => {
        this.$nextTick(() => this.onRender());
    });
},
onRender() {
    // do stuff after livewire renders
},
```

**Reply by dspaccapeli** [reply]
It seems like the `$wire.on` and `$wire.watch` methods get triggered both before the DOM is updated. Using Alpine's $nextTick works, but feels a bit 'hack-y'. I agree it would be nice if there would be another mechanism.

It could be achieved with Livewire morph hooks, but then the function would get triggered at every morph event. Doing it with $wire would feel cleaner/more efficient imo.

(EDIT: The following is not needed)
~~By the way you don't need to use the `rendered()` function, it should have an extra round-trip.
You can achieve the same by using inside of your client side $wire event listener:~~

```
$wire.on('event-name', () => {
   $nextTick(() => yourFunctionAfterRender())
   }
);
```

**Reply by aoxborrow** [reply]
> By the way you don't need to use the `rendered()` function, it should have an extra round-trip.
> You can achieve the same by using inside of your client side $wire event listener:

huh? there is no extra round-trip. the `rendered()` function is what is sending my custom `rendered` event. the `rendered()` PHP function is a built-in [Livewire lifecycle hook](https://livewire.laravel.com/docs/lifecycle-hooks#render) so it gets called automatically. btw, I'm using `->self()` now to constrain these events to same component only 
```php
    public function rendered(): void {
        // event so JS components can listen for re-render
        $this->dispatch('rendered')->self();
    }
```

**Reply by dspaccapeli** [reply]
You’re right — I misinterpreted how you were using the `rendered()` function and the way you were handling the event dispatch.

I do something similar, but call the custom event whenever the variable, for which I need to wait for the DOM to be re-rendered, changes.

**Reply by brandopnw** [reply]
Thanks so much.  This was really helpful
